crates/bevy_ecs/src/schedule/run_criteria.rs

```diff
@@ -111,59 +106,154 @@ impl RunCriteriaContainer {
     pub fn from_descriptor(descriptor: RunCriteriaDescriptor) -> Self {
         Self {
             should_run: ShouldRun::Yes,
-            inner: match descriptor.system {
-                RunCriteriaSystem::Single(system) => RunCriteriaInner::Single(system),
-                RunCriteriaSystem::Piped(system) => RunCriteriaInner::Piped { input: 0, system },
-            },
+            inner: descriptor.system,
+            parents: vec![],
             label: descriptor.label,
             before: descriptor.before,
             after: descriptor.after,
             archetype_generation: ArchetypeGeneration::initial(),
         }
     }
 
-    pub fn name(&self) -> Cow<'static, str> {
-        match &self.inner {
-            RunCriteriaInner::Single(system) => system.name(),
-            RunCriteriaInner::Piped { system, .. } => system.name(),
-        }
+    pub fn update_parent_indices(&mut self, labels: &HashMap<BoxedRunCriteriaLabel, usize>) {
+        self.parents = self
+            .after
+            .iter()
+            .take(self.inner.parents_len())
+            .map(|label| {
+                *labels.get(label).unwrap_or_else(|| {
```

Comment:
~~Nit: can't we just use `expect` here?~~

Nope, apparently this way dodges string formatting costs: https://stackoverflow.com/a/63291643/3234149